### PR TITLE
implement scheduling function with tests, modify Greeter class

### PR DESCRIPTION
**Pull request by danyag0820**
Implements query() in the FindMeetingQuery.java file, add test cases, modify Greet class to trim whitespace and special characters.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +272,115 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  public void optionalBusyPartOfDay() {
```
Comment (ankitamitra) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:300]: Can you add a @Test above this test? The "actual" seems to be missing the first possible TimeRange(?)

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 2]
```diff
@@ -30,4 +30,24 @@ public void testGreeting() {
...
+
+  @Test
+  public void testGreetingRemoveSymbols() {
+  Greeter greeter = new Greeter();
+
+  String greeting = greeter.greet("@$A#da!?");
+
+  // Symbols should be removed
+  Assert.assertEquals("Hello Ada", greeting);
+  }
```
Comment (ankitamitra) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:52]: nit: indentation (I think you need one level in).

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 3]
```diff
@@ -30,4 +30,24 @@ public void testGreeting() {
...
 }
```
Comment (aamanhey) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:53]: nit: newline

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,121 @@
...
+    for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (OPTIONAL_PEOPLE.contains(attendee)) {
+          optionsIncludingOptional =
+              updateAvailableTimes(optionsIncludingOptional, event.getWhen(), MEETING_DURATION);
+        }
+      }
+    }
+
+    if (optionsIncludingOptional.size() > 0 || MANDATORY_PEOPLE.size() == 0) {
```
Comment (ankitamitra) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:62]: Can use list.isEmpty() and !(list.isEmpty()) instead of size() > 0 and size() == 0

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java [hunk 5]
```diff
@@ -22,6 +22,6 @@
    * Returns a greeting for the given name.
    */
   public String greet(String name) {
-    return "Hello " + name;
+    return "Hello " + name.trim().replaceAll("[@#$%!?]","");
   }
 }
```
Comment (aamanhey) [inline walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java:27]: nit: newline

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,121 @@
...
+        }
+      } else {
+        if (time.duration() >= meetingDuration) {
+          proxyCurrentTimes.add(time);
+        }
+      }
+    }
+    return proxyCurrentTimes;
   }
 }
```
Comment (aamanhey) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:134]: With all of the if and else statements you might consider using a switch instead, it might be faster seeing as there are many different cases. Or consider doing a combination of if-else statements with a switch.

nit: newline

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,106 @@
...
+  }
+
+  /**
+   * Helper function that updates the List of available time ranges when a meeting can be scheduled
+   * @param currentTimes: The currently available time ranges
+   * @param unavailableTime: The time range when there is another event scheduled
+   * @param meetingDuration: The length of the meeting request
+   * @return A list of updated TimeRanges when a meeting can be scheduled
+   */
+  public List<TimeRange> updateAvailableTimes(List<TimeRange> currentTimes, TimeRange unavailableTime, long meetingDuration) {
```
Comment (cjwert-g) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:78]: nit: This line is too long and goes beyond what I can read in the Github code review UI. Can you use clang to format this?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,106 @@
...
+    for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (OPTIONAL_PEOPLE.contains(attendee)) {
+          optionsIncludingOptional =
+              updateAvailableTimes(optionsIncludingOptional, event.getWhen(), MEETING_DURATION);
+        }
+      }
+    }
+
+    if (!(optionsIncludingOptional.isEmpty()) || MANDATORY_PEOPLE.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (danyag0820) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:61]: My thought process is that MANDATORY_PEOPLE.isEmpty(), by checking if there are or aren't mandatory attendees, helps decide which time ranges to return. In the case of onlyOptionalGaps(), optionsIncludingOptional is filled(non-empty), so MANDATORY_PEOPLE.isEmpty() isn't necessary because either way we'll return the optional list.

However, in the case of noOptionalGaps(), when there are no mandatory attendees and optional attendees are not available, optionsIncludingOptional will be empty, but we still can't return options because options, by default, returns the time range for the entire day. By checking for mandatory attendees, it ensures that optionsIncludingOptional is returned even if it is empty.

I also think the way I used parentheses might've been misleading, but I meant for the line to evaluate to:
 if (!(false) || true) => if(true)

[the remark this answers is not in the document]
Comment (cjwert-g) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:61]: Ha, I didn't read all the parentheses correctly (I thought we were negating the result of the enitre OR statement, not just the optionsIncludingOptional.isEmpty()). So, most of this can be disregarded. That said, I think the parentheses around opitonsIncludingOptional.isEmpty() are unnecessary and make it harder to read.

And good point! Your reasoning makes sense for checking mandatoryPeople.

[the remark this answers is not in the document]
Comment (cjwert-g) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:61]: Disregard, since I misread this earlier! :P

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,106 @@
...
 public final class FindMeetingQuery {
+  /**Method that get the time ranges where a meeting request's mandatory attendees, and, if posible,
+   * optional attendees, are available for the request's duration.
+   * @param events: Collection of events that have already been scheduled
+   * @param request: Meeting request, which includes its attendees and duration
+   * @return A collection of TimeRanges when the meeting can be held
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    final long MEETING_DURATION = request.getDuration();
```
[the remark this answers is not in the document]
Comment (cjwert-g) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:31]: Saurabh pointed out that these links don't work!

Here's the same section on Github: https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

[the remark this answers is not in the document]
Comment (danyag0820) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:31]: thank you!
